### PR TITLE
feat(triggers): email spam prevention — test-fire lockdown, hourly cap, one-click unsubscribe

### PR DESCRIPTION
**Pull request by 0xdeafcafe**
Implements ADR-031: three abuse/cost protections for trigger email, sharing one threat model — LangWatch-operated email as a spam/cost vector. Slack is exempt throughout (posts to the customer's own workspace, no third-party victim, no per-message cost).

Stacked on #4498 (`feat/trigger-outbox-dispatch`); retarget to `main` after that merges.

### 1. Test fire sends only to the requester
- `testFireTemplate` no longer accepts a recipient list — the server resolves the recipient as the session user's email, so there is nothing to validate or trust. Slack webhook test fires unchanged.
- Per-user rate limit (`testfire:{userId}`, 10/min) with a retry-in-N-seconds error message.
- The drawer's test affordance shows "A test will be sent to *you@…*" instead of a recipients input.

### 2. Per-trigger hourly email cap
- `TRIGGER_EMAIL_HOURLY_CAP` (default 100) enforced in the outbox dispatcher's cadence stage: fixed-hour Redis counter (`trigger-email-cap:{projectId}:{triggerId}:{hourBucket}`), in-memory fallback when Redis is unconfigured **or errors mid-flight** (the cap must not fail open into the retry loop during a storm).
- Over the cap: the dispatch is dropped with `logger.error` (naming the trigger, suggesting a digest cadence), the send claim is still recorded so replays no-op, and the job completes — drops are terminal, not retried.
- Digest cadences (≤12 sends/hour) can never hit the cap; only `immediate` triggers do, by design.

### 3. One-click unsubscribe + suppression list
- New `EmailSuppression` model (projectId, lowercased email, nullable triggerId for project-wide scope) + migration, service/repository layers (the tRPC router is transport-only — name lookups and row enrichment live on `EmailSuppressionService`).
- Every trigger email gets a footer appended **outside** the customer's Liquid template (authors can't strip it) with two HMAC-signed links — "this notification only" / "all from this project" — plus `List-Unsubscribe` and `List-Unsubscribe-Post: One-Click` headers (Gmail/Yahoo bulk-sender requirements).
- Trigger emails switch from single-BCC to **per-recipient envelopes** (a one-click link must identify the recipient). The fan-out is idempotent across outbox retries: each delivery is recorded in the `TriggerSent` claim store under a `rcpt:`-prefixed key, so a partial provider failure retries only the unfinished recipients.
- SES raw-MIME sends (forced by the unsubscribe headers) are properly encoded: RFC 2047 subject, base64 body with RFC 2045 line folding.
- Dispatch-time filtering: suppressed recipients are skipped (filter runs before the cap so all-suppressed dispatches don't burn cap slots); the legacy custom-graph cron email path got the same filter + cap so its unsubscribe links aren't dead.
- Public `/unsubscribe` page (registered in the SPA router, works logged-out, masked email, two scope buttons) + RFC 8058 one-click endpoint as a Hono route at `POST /api/unsubscribe` — the `List-Unsubscribe` header targets the API endpoint while footer links open the page. Both IP-rate-limited; tampered tokens rejected via constant-time HMAC verify; empty `NEXTAUTH_SECRET` now throws instead of minting forgeable tokens.
- Project settings → **Email suppressions**: list (address, scope, date) + remove, gated `triggers:view`/`triggers:manage`.

## Docs
- `dev/docs/adr/031-trigger-email-abuse-protections.md` — decisions + rejected alternatives (auto-degrade-to-digest, auto-disable, verification-before-send, anonymous unsubscribe page).
- `specs/automations/spam-prevention.feature` — 18 scenarios; all covered by the implementation (verified scenario-by-scenario in review).

## Test plan
- Unit tests across the touched areas pass locally (dispatcher cap/suppression ordering, per-recipient retry idempotency + claim-key wiring, token round-trip/tamper, per-recipient fan-out + footer + headers, raw-MIME encoding incl. Unicode and long-line folding, one-click POST route 405/400/429/200/500, cap boundary + Redis-failure fallback, suppression service filtering + enrichment).
- Not exercised locally (no stack running): the Prisma migration against a real DB; manual click-through of `/unsubscribe` and the settings page.

## Known follow-ups (deliberate)
- No `.max()` on the trigger recipients list — recipient fan-out multiplies provider calls within the cap; pending a product call on the limit.
- Unsubscribe tokens don't expire — links in old emails keep working; TTL is a one-line add if preferred.
- No bounce ingestion yet (the suppression `reason` column is ready for it).

## Deployment Impact
- **New env var:** `TRIGGER_EMAIL_HOURLY_CAP` (optional, default `100`) — per-trigger hourly email dispatch cap. No action needed unless overriding.
- **New Prisma migration:** `20260611120000_create_email_suppression` creates the `EmailSuppression` table (single CREATE TABLE + indexes; instant, no backfill, no downtime).
- **Behavioural:** trigger emails switch from one BCC envelope to one send per recipient — provider call volume scales with recipient-list size; `NEXTAUTH_SECRET` must be set (already required by auth) or unsubscribe-token signing throws at first use.
- **No chart/service topology changes**; web + worker both pick the changes up on normal deploy.

## Summary by CodeRabbit

* **New Features**
  * Added email suppression management in Settings to view and remove suppressed email addresses
  * Added one-click unsubscribe links and RFC 8058-compliant headers to trigger emails
  * Added public unsubscribe page for recipients to manage notification preferences without login
  * Email test sends now restricted to the requesting user's inbox only
  * Implemented per-trigger hourly email dispatch limit to prevent notification storms
  * Added per-user rate limiting on test fire requests

### CODE REVIEW

File: dev/docs/adr/031-trigger-email-abuse-protections.md [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+### 2. Per-trigger hourly hard cap on dispatched emails
+
+In the outbox dispatcher's cadence stage (`handleCadenceBatch`), the email
+branch consults a Redis fixed-hour counter before sending:
+
+```
+key:    trigger-email-cap:{projectId}:{triggerId}:{floor(now / 1h)}
+INCR + EXPIRE 2h; if count > cap → drop
+```
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply dev/docs/adr/031-trigger-email-abuse-protections.md]: Addressed in 42860f6ad: language tag added to the fenced block.

File: langwatch/prisma/migrations/20260611120000_create_email_suppression/migration.sql [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    "email" TEXT NOT NULL,
+    "triggerId" TEXT,
+    "reason" TEXT NOT NULL DEFAULT 'unsubscribe',
+    "createdAt" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
+
+    CONSTRAINT "EmailSuppression_pkey" PRIMARY KEY ("id")
+);
+
+-- CreateIndex
+CREATE UNIQUE INDEX "EmailSuppression_projectId_email_triggerId_key" ON "EmailSuppression"("projectId", "email", "triggerId");
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/prisma/migrations/20260611120000_create_email_suppression/migration.sql]: Addressed in 42860f6ad: replaced the single composite unique with two partial unique indexes — one WHERE triggerId IS NOT NULL, one on (projectId, email) WHERE triggerId IS NULL — so project-wide rows dedupe correctly.

File: langwatch/prisma/schema.prisma [hunk 3]
```diff
@@ -1028,6 +1029,23 @@ model TriggerSent {
...
+  id        String   @id @default(nanoid())
+  projectId String
+  project   Project  @relation(fields: [projectId], references: [id])
+  email     String
+  triggerId String? // null = suppressed for every trigger in the project
+  reason    String   @default("unsubscribe")
+  createdAt DateTime @default(now())
+
+  @@unique([projectId, email, triggerId])
+  @@index([projectId])
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/prisma/schema.prisma:1050]: Addressed in 42860f6ad: schema keeps the @@unique for Prisma's client types with a comment noting the partial indexes live in the migration (Prisma can't express partial indexes natively).

File: langwatch/src/components/SettingsLayout.tsx [hunk 4]
```diff
@@ -161,9 +161,10 @@ export default function SettingsLayout({
 
           <NavSection
             label="Features"
-            paths={["/settings/annotation-scores", "/settings/topic-clustering", "/settings/data-retention"]}
+            paths={["/settings/annotation-scores", "/settings/topic-clustering", "/settings/data-retention", "/settings/email-suppressions"]}
           >
             <MenuLink href="/settings/data-retention">Data Retention</MenuLink>
+            <MenuLink href="/settings/email-suppressions">Email Suppressions</MenuLink>
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/components/SettingsLayout.tsx]: Addressed in 42860f6ad: nav entry now gated behind hasPermission("triggers:view"), same pattern as the Audit Log entry.

File: langwatch/src/pages/api/unsubscribe.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+const logger = createLogger("langwatch:unsubscribe:one-click");
+
+/**
+ * ADR-031: RFC 8058 one-click unsubscribe endpoint. Mail clients POST here
+ * (body `List-Unsubscribe=One-Click`) when the recipient hits the native
+ * "unsubscribe" affordance. The token in `?token=` is the authorization — its
+ * HMAC binds it to one recipient — so this route needs no session. One-click
+ * is trigger-scoped (the link the `List-Unsubscribe` header carries). Always
+ * returns 200 to a valid token so the mail client shows success; only a
+ * malformed/missing token is a 4xx.
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/pages/api/unsubscribe.ts]: Addressed in 42860f6ad: comment corrected to match the actual behavior (400 malformed token, 405 non-POST, 429 rate-limited).

File: langwatch/src/pages/settings/email-suppressions.tsx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+function EmailSuppressionsSettings() {
+  const { project } = useOrganizationTeamProject();
+  if (!project) return null;
+  return <EmailSuppressionsPage projectId={project.id} />;
+}
+
+export default withPermissionGuard("triggers:view", {
+  layoutComponent: SettingsLayout,
+})(EmailSuppressionsSettings);
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/pages/settings/email-suppressions.tsx:33]: Addressed in 42860f6ad: Remove button only renders with triggers:manage.

File: langwatch/src/pages/unsubscribe.tsx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        bg="bg.panel"
+        borderWidth="1px"
+        borderColor="border"
+        borderRadius="lg"
+        padding={8}
+        maxW="480px"
+        width="full"
+      >
+        {!token || resolved.isError ? (
+          <VStack align="start" gap={2}>
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/pages/unsubscribe.tsx:45]: Addressed in 42860f6ad: spinner shown until router.isReady, so the invalid-link state can't flash during hydration.

File: langwatch/src/server/api/routers/__tests__/rateLimitMessage.unit.test.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    it("clamps the retry to at least one second", () => {
+      const now = 1_000_000;
+      const message = buildRetryAfterMessage({
+        prefix: "Too many test fires.",
+        resetAt: now - 5_000,
+        now,
+      });
+      expect(message).toBe("Too many test fires. Try again in 1 second.");
+    });
+  });
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/api/routers/__tests__/rateLimitMessage.unit.test.ts:45]: Addressed in 42860f6ad: added the when-level describe nesting.

File: langwatch/src/server/api/routers/automations.ts [hunk 9]
```diff
@@ -427,17 +428,47 @@ export const automationRouter = createTRPCRouter({
...
+      });
+      if (!limit.allowed) {
+        throw new TRPCError({
+          code: "TOO_MANY_REQUESTS",
+          message: buildRetryAfterMessage({
+            prefix: "Too many test fires.",
+            resetAt: limit.resetAt,
+          }),
+        });
+      }
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/api/routers/automations.ts]: Addressed in 42860f6ad: rate limit now scoped to the email channel only — Slack test fires hit the customer's own webhook, so per ADR-031 the limit is mail-provider hygiene and doesn't apply.

File: langwatch/src/server/api/routers/emailSuppression.ts [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+      const nameById = new Map(triggers.map((t) => [t.id, t.name]));
+      return rows.map((r) => ({
+        id: r.id,
+        email: r.email,
+        triggerId: r.triggerId,
+        triggerName: r.triggerId != null ? nameById.get(r.triggerId) ?? null : null,
+        reason: r.reason,
+        createdAt: r.createdAt,
+      }));
+    }),
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/api/routers/emailSuppression.ts:140]: Addressed in 42860f6ad: both getAll and remove now write audit entries via the existing auditLog pattern (record counts and ids in args, no raw emails).

File: langwatch/src/server/app-layer/triggers/repositories/emailSuppression.prisma.repository.ts [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    // is matched/created via findFirst + create instead of upsert. Both paths
+    // are idempotent: an existing row is returned untouched.
+    if (triggerId === null) {
+      const existing = await this.prisma.emailSuppression.findFirst({
+        where: { projectId, email, triggerId: null },
+      });
+      if (existing) return existing;
+      return this.prisma.emailSuppression.create({
+        data: { projectId, email, triggerId: null, reason },
+      });
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/app-layer/triggers/repositories/emailSuppression.prisma.repository.ts:67]: Addressed in 42860f6ad: with the new partial unique index on (projectId, email) WHERE triggerId IS NULL, the null-trigger path now creates and catches P2002 → re-fetches the existing row. Idempotent under concurrency, no findFirst+create gap.

File: langwatch/src/server/app-layer/triggers/repositories/emailSuppression.repository.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    reason: string;
+  }): Promise<EmailSuppressionRow> {
+    return {
+      id: "null-suppression",
+      projectId: params.projectId,
+      email: params.email,
+      triggerId: params.triggerId,
+      reason: params.reason,
+      createdAt: new Date(),
+    };
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/app-layer/triggers/repositories/emailSuppression.repository.ts:92]: Addressed in a6872c258: the null repo now lowercases/trims the returned email, matching the documented contract.

File: langwatch/src/server/event-sourcing/outbox/dispatcher.ts [hunk 13]
```diff
@@ -78,6 +78,30 @@ export interface OutboxDispatcherDeps {
...
+   * ADR-031: per-trigger hourly hard cap on dispatched emails. Consults a
+   * fixed-hour Redis counter; `allowed: false` means the trigger has already
+   * sent `cap` emails this hour and this dispatch must be dropped (not sent,
+   * not retried). Injected so tests can fake it. Slack never calls it.
+   */
+  consumeEmailCapSlot: (args: {
+    projectId: string;
+    triggerId: string;
+    now: Date;
+  }) => Promise<{ allowed: boolean; count: number }>;
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/event-sourcing/outbox/dispatcher.ts:92]: Keeping the non-idempotent reservation deliberately: a retry can re-INCR the hourly counter, but the worst case is the cap tripping slightly early (fewer emails) — never extra sends or double-delivery. The cap is a fixed-window approximation by design (see ADR-031), and a per-dispatch SETNX reservation adds a Redis key shape + cleanup path for no safety gain.

File: langwatch/src/server/event-sourcing/outbox/dispatcher.ts [hunk 14]
```diff
@@ -354,7 +378,49 @@ async function handleCadenceBatch(
...
+              projectId,
+              triggerId,
+              count: capSlot.count,
+              cap: deps.emailHourlyCap,
+            },
+            "Trigger exceeded its hourly email cap — dropping this dispatch. " +
+              "Switch this trigger to a digest cadence to coalesce its volume.",
+          );
+          break;
+        }
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/event-sourcing/outbox/dispatcher.ts:431]: Addressed in 42860f6ad: drops no longer run success-path bookkeeping — updateLastRunAt and the dispatched log are skipped when nothing was sent, with a distinct 'claimed but not sent' log instead. claimSend still runs on drops (ADR-031 mandates replays no-op).

File: langwatch/src/server/event-sourcing/outbox/emailHourlyCap.ts [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+}): Promise<{ allowed: boolean; count: number }> {
+  const hourBucket = Math.floor(now.getTime() / HOUR_MS);
+  const key = `trigger-email-cap:${projectId}:${triggerId}:${hourBucket}`;
+
+  if (connection) {
+    try {
+      const count = await connection.incr(key);
+      if (count === 1) {
+        await connection.expire(key, EXPIRE_SECONDS);
+      }
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/event-sourcing/outbox/emailHourlyCap.ts]: Addressed in 42860f6ad: EXPIRE now runs with NX on every hit instead of only count===1, so a missed first-hit expiry can no longer leak the key.

File: langwatch/src/server/mailer/__tests__/triggerEmail.unit.test.tsx [hunk 16]
```diff
@@ -49,16 +51,47 @@ describe("sendTriggerEmail", () => {
...
+        headers: Record<string, string>;
+      };
+      expect(args.html).toContain("Stop receiving this notification");
+      expect(args.html).toContain("Stop all notifications from this project");
+      expect(args.html).toContain("/unsubscribe?token=");
+      expect(args.headers["List-Unsubscribe"]).toMatch(/^<.*\/unsubscribe\?token=/);
+      expect(args.headers["List-Unsubscribe-Post"]).toBe(
+        "List-Unsubscribe=One-Click",
+      );
+    });
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/mailer/__tests__/triggerEmail.unit.test.tsx:198]: Addressed in 42860f6ad: the test now asserts one envelope per recipient with recipient-specific (distinct) unsubscribe tokens.

File: langwatch/src/server/mailer/__tests__/unsubscribe.read.unit.test.ts [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+describe("resolveUnsubscribe", () => {
+  let deps: { lookupNames: ReturnType<typeof vi.fn<ResolveDeps["lookupNames"]>> };
+
+  beforeEach(() => {
+    deps = { lookupNames: vi.fn<ResolveDeps["lookupNames"]>() };
+  });
+
+  describe("given a valid trigger-scoped token", () => {
+    it("returns masked email plus project and trigger names", async () => {
+      deps.lookupNames.mockResolvedValue({
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/mailer/__tests__/unsubscribe.read.unit.test.ts]: Addressed in 42860f6ad: restructured with when-level describes per the BDD house rules.

File: langwatch/src/server/mailer/__tests__/unsubscribeToken.unit.test.ts [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+} from "../unsubscribeToken";
+
+describe("unsubscribeToken", () => {
+  describe("given NEXTAUTH_SECRET is empty", () => {
+    afterEach(() => {
+      vi.restoreAllMocks();
+    });
+
+    describe("when signing or verifying", () => {
+      it("throws rather than minting forgeable tokens", () => {
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/mailer/__tests__/unsubscribeToken.unit.test.ts]: Addressed in 42860f6ad: describe wording narrowed to match what the block actually asserts.

File: langwatch/src/server/mailer/triggerEmail.tsx [hunk 19]
```diff
@@ -92,6 +146,65 @@ export const sendTriggerEmail = async ({
...
+      projectId,
+      triggerId,
+      email: recipient,
+      baseHost,
+    });
+    await sendEmail({
+      to: noReplyTo,
+      bcc: [recipient],
+      subject,
+      html: html + footerHtml,
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/mailer/triggerEmail.tsx]: Addressed in 42860f6ad: footer is now injected before the closing </body> tag when present (case-insensitive), appended otherwise — covered by unit tests for both branches.

File: langwatch/src/server/mailer/unsubscribe.read.ts [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+}): Promise<void> {
+  const payload = verifyUnsubscribeToken(token);
+  if (!payload) {
+    throw new InvalidUnsubscribeTokenError();
+  }
+  await deps.suppress({
+    projectId: payload.projectId,
+    email: payload.email,
+    triggerId: scope === "project" ? null : payload.triggerId,
+  });
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/mailer/unsubscribe.read.ts]: Working as intended: the token binds projectId + email via HMAC, and both scopes only ever suppress the token's own email within the token's own project. The scope choice ('this notification' vs 'whole project') is deliberately the recipient's — the footer offers both, and a trigger-scoped token escalating to project-wide suppression of the same address grants no privilege the recipient doesn't already have. Binding scope into the token would just mean minting two tokens per footer for no security gain.

File: langwatch/src/server/mailer/unsubscribeToken.ts [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+ * Wire format: `base64url(JSON payload) + "." + hex(HMAC of the JSON payload)`.
+ */
+
+export interface UnsubscribePayload {
+  projectId: string;
+  /** null = suppress every trigger in the project. */
+  triggerId: string | null;
+  /** Always lowercased before signing and on verify, so a link works
+   *  regardless of the recipient address casing the author typed. */
+  email: string;
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/mailer/unsubscribeToken.ts:22]: Addressed in 42860f6ad: verifyUnsubscribeToken now normalizes (lowercase/trim) the decoded email, matching the sign side and the doc comment.

File: langwatch/src/server/mailer/unsubscribeToken.ts [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+  const value = env.NEXTAUTH_SECRET ?? "";
+  if (!value) {
+    // An empty secret makes tokens forgeable — anyone could mint a valid
+    // unsubscribe link for any address. Fail closed rather than sign/verify
+    // with an empty key.
+    throw new Error(
+      "NEXTAUTH_SECRET is not set; refusing to sign/verify unsubscribe tokens with an empty key.",
+    );
+  }
+  return value;
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/mailer/unsubscribeToken.ts:35]: Keeping NEXTAUTH_SECRET for now: it matches the established keyed-hash pattern in triggerNoReply.ts, and the failure mode of a secret rotation is bounded (old unsubscribe links stop verifying; new emails carry fresh links — recipients can still unsubscribe from any newer email). A dedicated signing secret is a reasonable follow-up if secret rotation becomes routine, but it adds an env var + migration story this PR doesn't need.

File: specs/automations/spam-prevention.feature [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+  ceiling, and external recipients have no way to opt out. Three protections
+  close these, all email-only — Slack posts to the customer's own workspace
+  via a webhook they provisioned, so it carries no third-party victim and no
+  per-message cost.
+
+  See dev/docs/adr/031-trigger-email-abuse-protections.md.
+
+  Background:
+    Given a project with automations enabled
+
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply specs/automations/spam-prevention.feature:14]: Keeping the one-line declarative Background — it matches the repo's own spec conventions (dispatch-timing.feature, authoring-drawer.feature, and notification-templates.feature all open the same way).

**Comment by sergioestebance** [comment]
## Blocking review: not ready to merge

Revalidated against current head `a6872c2580ca46ba6ae6089dabeec2c6422c2219`.

### P1: RFC 8058 one-click targets the browser page, not the POST handler

[`triggerEmail.tsx:55-75`](https://github.com/langwatch/langwatch/blob/a6872c2580ca46ba6ae6089dabeec2c6422c2219/langwatch/src/server/mailer/triggerEmail.tsx#L55-L75) builds both footer links and `List-Unsubscribe` with `/unsubscribe?token=...`. The only POST handler that persists a suppression is [`pages/api/unsubscribe.ts`](https://github.com/langwatch/langwatch/blob/a6872c2580ca46ba6ae6089dabeec2c6422c2219/langwatch/src/pages/api/unsubscribe.ts), and there is no rewrite connecting the routes. A mail client's one-click POST therefore hits the page route and never calls `confirmUnsubscribe`.

Keep footer links on `/unsubscribe`, but use `/api/unsubscribe?token=...` for the header. Add a regression test that distinguishes the browser and one-click URLs.

### P1: The router bypasses the required app/service/repository layer

[`emailSuppression.ts:63-75`](https://github.com/langwatch/langwatch/blob/a6872c2580ca46ba6ae6089dabeec2c6422c2219/langwatch/src/server/api/routers/emailSuppression.ts#L63-L75) directly queries `ctx.prisma.project` and `ctx.prisma.trigger`; [`emailSuppression.ts:157-164`](https://github.com/langwatch/langwatch/blob/a6872c2580ca46ba6ae6089dabeec2c6422c2219/langwatch/src/server/api/routers/emailSuppression.ts#L157-L164) does another direct trigger query. This conflicts with the repository convention in `CLAUDE.md`: **“Routes call services only.”** It also splits suppression-list assembly across transport and persistence layers despite `getApp().emailSuppressions` already owning this domain.

Add focused service/repository methods for unsubscribe display metadata and enriched suppression rows. Keep the router limited to validation, authorization/rate limiting, service calls, auditing, and transport error mapping.

### P1: Per-recipient fan-out re-sends successful recipients on retry

[`triggerEmail.tsx:191-219`](https://github.com/langwatch/langwatch/blob/a6872c2580ca46ba6ae6089dabeec2c6422c2219/langwatch/src/server/mailer/triggerEmail.tsx#L191-L219) sends recipients sequentially. If recipient A succeeds and recipient B fails retryably, the function throws. The dispatcher only writes `TriggerSent` after the entire call succeeds, so the outbox retry runs the full loop and sends A again.

Persist recipient-level completion/idempotency, or otherwise retry only unfinished recipients. Add a test where the first provider call resolves, the second rejects, and retry does not call the first recipient again.

### P1: SES raw MIME emits arbitrary UTF-8 as unencoded `7bit`

[`emailSender.ts:109-125`](https://github.com/langwatch/langwatch/blob/a6872c2580ca46ba6ae6089dabeec2c6422c2219/langwatch/src/server/mailer/emailSender.ts#L109-L125) writes the raw `Subject` without RFC 2047 encoding and emits HTML unchanged with `Content-Transfer-Encoding: 7bit`. Adding unsubscribe headers forces every SES trigger email through this raw path. Trigger names and templates can contain non-ASCII text and long lines, so messages can be rejected or corrupted.

Use a MIME builder or correctly encode/fold the subject and body with quoted-printable/base64. Add tests for Unicode subject/body and long HTML lines. AWS requirements: https://docs.aws.amazon.com/ses/latest/dg/send-email-raw.html

File: langwatch/src/server/api/routers/emailSuppression.ts [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+      await enforceUnsubscribeRateLimit({
+        ip: getClientIp(ctx.req),
+        action: "resolve",
+        max: 30,
+      });
+      const view = await resolveUnsubscribe({
+        token: input.token,
+        deps: {
+          lookupNames: async ({ projectId, triggerId }) => {
+            const project = await ctx.prisma.project.findFirst({
```
Comment (sergioestebance) [inline langwatch/src/server/api/routers/emailSuppression.ts]: service here'd be nice

**Comment by 0xdeafcafe** [comment]
All four addressed in c593ca5d1:

**RFC 8058 one-click** — `List-Unsubscribe` now points at `/api/unsubscribe` while footer links stay on `/unsubscribe` (regression test asserts the two paths diverge). Beyond the URL: the POST handler was actually dead code (nothing mounted `pages/api/unsubscribe.ts` — `/api/*` is served by the Hono router only), and neither `/unsubscribe` nor `/settings/email-suppressions` was registered in `routes.tsx`, so all three surfaces 404'd. The handler is ported to a Hono route mounted at `POST /api/unsubscribe` (same 405/400/429/200/500 contract, declared `publicEndpoint`), and both pages are registered.

**Router layering** — `emailSuppression.ts` no longer touches Prisma: `resolveUnsubscribeToken` delegates to `EmailSuppressionService.lookupNames` and `getAll` uses `getAllEnriched` (rows arrive with `triggerName`), both backed by a new name-lookup repository wired through presets. Router keeps validation, authz, rate limiting, auditing, and error mapping only.

**Per-recipient retry idempotency** — `sendPerRecipient` now consults an `isRecipientSent`/`recordRecipientSent` gate (recipient hash, recorded only after a successful provider call). The dispatcher backs it with the existing `TriggerSent` claim store under a `rcpt:<dispatch-digest>:<hash>` key, stable across retries of the same dispatch. Tests cover both the mailer behaviour (first recipient succeeds, second fails, retry skips the first) and the dispatcher wiring (read and write target the same key).

**SES raw MIME** — Subject is RFC 2047-encoded when non-ASCII, HTML body ships base64 with RFC 2045 76-char folding (attachment base64 folded too). New `emailSender.unit.test.ts` covers Unicode subject/body, long-line folding, header placement before the body separator, CRLF stripping, and raw-vs-simple command selection.

File: langwatch/src/server/api/routers/emailSuppression.ts (hunk 24 above)
Comment (0xdeafcafe) [reply langwatch/src/server/api/routers/emailSuppression.ts]: Addressed in c593ca5d1: the Prisma lookups moved into EmailSuppressionService.lookupNames / getAllEnriched (backed by a name-lookup repository, wired via presets); the router now does auth/validation/rate-limit/audit/error-mapping only.

Comment (0xdeafcafe) [reply langwatch/src/server/api/routers/emailSuppression.ts]: Went further in 9462e74df: unsubscribe.read.ts is gone — token resolve/confirm orchestration now lives on EmailSuppressionService itself (no deps-injection round-trip), and both transports (tRPC router + one-click Hono route) are pure mapping: validation, rate limit, audit, service call, error translation.
